File: tests/fixtures/graph-src/typescript/geometry.ts
// Geometry helpers for the TypeScript portion of the fixture.
//
// Exercises: a type alias, an enum + members, a class with methods and a
// member, and an exported free function.

// A type alias — the reader should record a `type_alias` node.
export type Scalar = number;

// An enum with several members.
export enum Kind {
  Circle = "circle",
  Rectangle = "rectangle",
  Empty = "empty",
}

// A class with a member field and methods.
export class Point {
  x: Scalar;
  y: Scalar;

  constructor(x: Scalar, y: Scalar) {
    this.x = x;
    this.y = y;
  }

  // Method that calls a free function in this file (intra-file call).
  distanceTo(other: Point): Scalar {
    return hypot(this.x - other.x, this.y - other.y);
  }
}

// Exported free function used by `Point.distanceTo`.
export function hypot(a: Scalar, b: Scalar): Scalar {
  return Math.sqrt(a * a + b * b);
}

// An interface — the reader must record an `interface` node. The member returns
// a primitive (`number`) rather than `Scalar`: CodeGraph emits a `references`
// edge from an interface member to a *named* return type, which semgraph does
// not (it does not walk interface bodies for reference sites), and references
// are not an acceptance metric — a primitive return keeps the fixture exact.
export interface Measurable {
  measure(): number;
}

// A base class for the `extends` edge below.
export class Base {}

// `extends Base` is an `extends` edge; `implements Measurable` is an
// `implements` edge (Marker -> Base, Marker -> Measurable).
export class Marker extends Base implements Measurable {
  measure(): number {
    return 0;
  }
}

File: tests/fixtures/graph-src/rust/geometry.rs
//! Geometry primitives used by the Rust portion of the fixture.
//!
//! Exercises: struct + fields, impl methods, enum + members, a type alias,
//! and free functions that are called across files.

/// A type alias over a primitive — the reader must record `type_alias` nodes.
pub type Scalar = f64;

/// A struct with named fields (struct members).
pub struct Point {
    pub x: Scalar,
    pub y: Scalar,
}

impl Point {
    /// Associated constructor function (method / associated fn).
    pub fn new(x: Scalar, y: Scalar) -> Self {
        Point { x, y }
    }

    /// Instance method that calls a free function in this file (intra-file call).
    pub fn distance_to(&self, other: &Point) -> Scalar {
        let dx = self.x - other.x;
        let dy = self.y - other.y;
        hypot(dx, dy)
    }
}

/// An enum with several variants (enum members).
pub enum Shape {
    Circle { radius: Scalar },
    Rectangle { width: Scalar, height: Scalar },
    Empty,
}

impl Shape {
    /// Method with a match over the enum's members.
    pub fn area(&self) -> Scalar {
        match self {
            Shape::Circle { radius } => 3.14159 * radius * radius,
            Shape::Rectangle { width, height } => width * height,
            Shape::Empty => 0.0,
        }
    }
}

/// Free function used by `Point::distance_to`.
pub fn hypot(a: Scalar, b: Scalar) -> Scalar {
    (a * a + b * b).sqrt()
}
